// @airgap/beacon-transport-walletconnect/dist/esm/WalletConnectTransport.js
import { WalletConnectCommunicationClient } from './communication-client/WalletConnectCommunicationClient';
import { Origin, TransportStatus } from '@airgap/beacon-types';
import { Transport, PeerManager } from '@airgap/beacon-core';
/**
 * @internalapi
 *
 *
 */
export class WalletConnectTransport extends Transport {
    wcOptions;
    // public readonly type: TransportType = TransportType.WALLETCONNECT
    constructor(name, _keyPair, storage, storageKey, wcOptions) {
        super(name, WalletConnectCommunicationClient.getInstance(wcOptions), new PeerManager(storage, storageKey));
        this.wcOptions = wcOptions;
    }
    static async isAvailable() {
        return Promise.resolve(true);
    }
    async connect() {
        if (this._isConnected !== TransportStatus.NOT_CONNECTED) {
            return;
        }
        this._isConnected = TransportStatus.CONNECTING;
        await this.client.init();
        const knownPeers = await this.getPeers();
        if (knownPeers.length > 0) {
            knownPeers.map(async (peer) => this.listen(peer.publicKey));
        }
        await this.startOpenChannelListener();
        return super.connect();
    }
    async hasPairings() {
        return (await this.client.storage.hasPairings())
            ? true
            : !!this.client.signClient?.pairing.getAll()?.length;
    }
    async hasSessions() {
        return (await this.client.storage.hasSessions())
            ? true
            : !!this.client.signClient?.session.getAll()?.length;
    }
    async closeActiveSession(account) {
        if (!(await this.hasPairings()) || !(await this.hasPairings())) {
            await this.disconnect();
        }
        else {
            await this.client.closeActiveSession(account.address);
        }
    }
    async getPeers() {
        const client = WalletConnectCommunicationClient.getInstance(this.wcOptions);
        const session = client.currentSession();
        if (!session) {
            return [];
        }
        return [
            {
                senderId: session.peer.publicKey,
                extensionId: session.peer.metadata.name,
                id: session.peer.publicKey,
                type: 'walletconnect-pairing-response',
                name: 'peer',
                publicKey: session.peer.publicKey,
                version: 'first'
            }
        ];
    }
    async disconnect() {
        await this.client.close();
        return super.disconnect();
    }
    async startOpenChannelListener() {
        //
    }
    getPairingRequestInfo() {
        return this.client.getPairingRequestInfo();
    }
    async listen(publicKey) {
        await this.client
            .listenForEncryptedMessage(publicKey, (message) => {
            const connectionContext = {
                origin: Origin.WALLETCONNECT,
                id: publicKey
            };
            this.notifyListeners(message, connectionContext).catch((error) => {
                throw error;
            });
        })
            .catch((error) => {
            throw error;
        });
    }
}
